### ISSUE TITLE
REPL doesn't auto-complete re-exported symbols

### ISSUE THREAD

**Issue by rndnoise** [purescript/purescript#3227]
Given a module like this:

```purescript
module X ( foo, module Data.Maybe ) where
import Data.Maybe (maybe)

foo :: String
foo = "ok"
```

When running `pulp repl`, and entering `import X as X`, then typing `X.[tab]` will only show `foo` but not `maybe`. This also happens when importing `X` unqualified.

**Comment by rndnoise** [comment]
I would be glad to work on this if a pull request would be accepted!

**Comment by LiamGoodacre** [comment]
Thanks for reporting.  A pull request for this would be most welcomed.

**Comment by rndnoise** [comment]
This is ready for review. I ended up making more changes than fixing re-exports as I noticed some other related problems (at least, I think they are problems).

* If the previous word is `::`, the REPL now suggests type names
* The REPL previously would suggest fully-qualified names that haven't been imported (but were loaded modules). Choosing these would cause an "unknown module" or "unknown value".
* Previously, values and types defined within the REPL session were not offered as suggestions, but are now.
* The `:kind` and `:type` commands only offered to complete one more word, but now will offer completions of unlimited words (helpful for multi-word expressions like `:type map ...` or `:kind Array ...`)

Rather than recompute the list of names each time a `completion'` is run, I chose to add two fields to `PSCiState` which are only recomputed after a module is imported, a let-binding is added, or the `:clear` or `:reload` commands are given. If this change is too intrusive, it should be simple to move the stuff in `updateImportsExports` back into the `Completions.hs` module.

These changes broke many tests, but only things that directly tested the tab completions. So I'm not certain if there's something that depends on the old behavior (eg, completing fully-qualified names that haven't been imported).

I'll be happy to fix the broken tests as long as doing so will move this PR along to be accepted.

**Comment by hdgarrood** [comment]
This looks really promising! The approach looks good. To get this to a mergeable state I'd like for it to include a bunch of new tests which would have previously failed, too.

> * If the previous word is ::, the REPL now suggests type names
> * The :kind and :type commands only offered to complete one more word, but now will offer completions of unlimited words (helpful for multi-word expressions like :type map ... or :kind Array ...)
> * Don't include '[' or '(' prefixes as part of the name to complete

These are all great ideas, but I'd rather consider them separately.

**Comment by hdgarrood** [comment]
I've just realised this is a duplicate of #2080. @garyb had some ideas about how to go about this, so actually I'd like to check with him before committing to this approach.

**Comment by garyb** [comment]
😬 I hate that there are so many issues that end with me being like "yeah, I'm about to take a look at this" and then die for two years. Ahem.

I think any steps we can take now aside from the thing I was talking about will still be good to have. If/when I do eventually get around to doing what I said in that issue it might help simplify some bits, but it's not worth holding things up for.

**Comment by rndnoise** [comment]
OK, I'll separate the extra stuff into a different request and start updating the tests. Thanks for the quick feedback @garyb and @hdgarrood.

**Comment by rndnoise** [comment]
Ready for another review: I've updated the tests, adding some that wouldn't have passed previously and updating some that failed as a result of the changes. I also removed the changes that @hdgarrood asked to consider separately.

**Comment by rndnoise** [comment]
I think all the feedback was addressed (in #3231), but let me know if any more work is needed.
